beacon_node/beacon_chain/src/attestation_rewards.rs

```diff
@@ -29,18 +38,85 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
         debug!(log, "computing attestation rewards"; "epoch" => epoch, "validator_count" => validators.len());
 
         // Get state
-        let spec = &self.spec;
-
         let state_slot = (epoch + 1).end_slot(T::EthSpec::slots_per_epoch());
 
         let state_root = self
             .state_root_at_slot(state_slot)?
             .ok_or(BeaconChainError::NoStateForSlot(state_slot))?;
 
-        let mut state = self
+        let state = self
             .get_state(&state_root, Some(state_slot))?
             .ok_or(BeaconChainError::MissingBeaconState(state_root))?;
 
+        match state {
+            BeaconState::Base(_) => self.compute_attestation_rewards_base(state, validators),
+            BeaconState::Altair(_) | BeaconState::Merge(_) | BeaconState::Capella(_) => {
+                self.compute_attestation_rewards_altair(state, validators)
+            }
+        }
+    }
+
+    fn compute_attestation_rewards_base(
+        &self,
+        mut state: BeaconState<T::EthSpec>,
+        validators: Vec<ValidatorId>,
+    ) -> Result<StandardAttestationRewards, BeaconChainError> {
+        let spec = &self.spec;
+        let mut validator_statuses = ValidatorStatuses::new(&state, spec)?;
+        validator_statuses.process_attestations(&state)?;
+
+        let ideal_rewards =
+            self.compute_ideal_rewards_base(&state, &validator_statuses.total_balances)?;
+
+        let validator_indices = Self::validators_ids_to_indices(&mut state, validators)?;
+        let indices_to_attestation_delta = get_attestation_deltas_subset(
+            &state,
+            &validator_statuses,
+            Some(&validator_indices),
+            spec,
+        )?;
+
+        let mut total_rewards = vec![];
+
+        for (index, delta) in indices_to_attestation_delta.into_iter() {
```

Comment:
If we make the change I suggested to revert `get_attestation_deltas`, we can do the filtering by index here, with an `.enumerate()` to get the indices.

Reply:
I haven't reverted `get_attestation_deltas`, but I've added the filtering logic to `get_attestation_deltas_subset`.

so now we have:
- `pub fn get_attestation_deltas_all`: this calls into `get_attestation_deltas` without specifying a subset, code path should be almost identical to the previous `get_attestation_deltas`
- `pub fn get_attestation_deltas_subset`: this calls into `get_attestation_deltas` with a specified subset, and filter the results to only include the subset
- `fn get_attestation_deltas`: this is where the delta computation logic is, I've left this function as private because the results need to be filtered for the subset scenario, which is done in `get_attestation_deltas_subset`.

Reply:
perfect